### DISCUSSION TITLE
JFreechart 1.0.19 Data Area size

### DISCUSSION THREAD

**Discussion by sonideft** [jfree/jfreechart#326] [Q&A]
I am looking to find out the exact size in pixels of the Data area of an XYPlot. I know version 1.5x has a function you can call for this but I can't move to that version just yet. The end goal is related to scaling the draw result on a custom Graphics driver (DXFGraphics via https://jsevy.com). The approach is scale the Graphics instance before drawing so that the final output matches the real values plotted (in meters or feet) as opened in AutoCad and measured.

I've looked through the documentation and have actually gotten quite far with accomplishing this, but there is some padding or something that is not being considered.

Here is the relevant section of code that works perfectly for vertical scaling and extremely close for horizontal. Please help me find the missing padding I'm missing or misapplying. Thanks.

double hfactor=RefData.getDefault().getDxf_hscale_factor();//usually 2.0, as selected by user
        RectangleInsets padding = this.getChart().getPadding();
        RectangleInsets insets = this.getChart().getPlot().getInsets();
        RectangleInsets axisoffsets = this.getChart().getXYPlot().getAxisOffset();
        ValueAxis domainaxis=this.getChart().getXYPlot().getDomainAxis();
        ValueAxis rangeaxis=this.getChart().getXYPlot().getRangeAxis();
        Rectangle visiblerect=this.getVisibleRect();//incoming visible rectangle starts at 0,0

        double dataAreaXLeft=visiblerect.x + padding.getLeft()+insets.getLeft()+axisoffsets.getLeft();
        double dataAreaBottom=visiblerect.y + padding.getBottom()+insets.getBottom()+axisoffsets.getBottom();
        double dataAreaXRight=visiblerect.getMaxX() - padding.getRight() - insets.getRight() - axisoffsets.getRight();
        double dataAreaTop=visiblerect.getMaxY() - padding.getTop() - insets.getTop() - axisoffsets.getTop();
        double adjustedheight=dataAreaTop-dataAreaBottom-17.0;//fudge factor
        double adjustedwidth=dataAreaXRight-dataAreaXLeft-20.0;//fudge factor
        this.getChart().getPlot().setBackgroundPaint(null);
        this.getChart().setBackgroundPaint(null);
        this.setBackground(null);
        this.getChart().fireChartChanged();
        
        // get the DXFGraphics object to draw into
        DXFGraphics dxfGraphics = dxfDocument.getGraphics();
        double vscale = rangeaxis.getRange().getLength()/adjustedheight;
        double hscale = domainaxis.getRange().getLength()/(adjustedwidth*hfactor);
        dxfGraphics.scale(hscale, vscale);

        MyChartPanel.this.getChart().draw(dxfGraphics, MyChartPanel.this.getVisibleRect());

**Comment by trashgod** [comment]
You might look at an `AxisChangeListener`, seen [here](https://stackoverflow.com/a/13362809/230513).

**Reply by sonideft** [reply]
Thanks, but I don't think that helps me. In my case every pixel that is printed to the graphics device needs the correct horizontal and vertical scale to be drawn correctly in precise units for the engineering format it is going into.
